src/engine/Server.cpp

```diff
@@ -492,6 +494,139 @@ Awaitable<void> Server::process(
+  // resolve the variables using each row of the result to obtain a quad without
+  // variables. These quads can then be used with `DeltaTriples`.
+  auto resolveVariable = [](const IdTable& idTable, const size_t& row,
+                            const VariableToColumnMap& variableColumns,
+                            IdOrVariable idOrVar) -> std::optional<Id> {
+    if (std::holds_alternative<Variable>(idOrVar)) {
+      auto var = std::get<Variable>(std::move(idOrVar));
+      if (!variableColumns.contains(var)) {
+        return std::nullopt;
+      } else {
+        return idTable(row, variableColumns.at(var).columnIndex_);
+      }
+    } else if (std::holds_alternative<Id>(idOrVar)) {
+      return std::get<Id>(idOrVar);
+    } else {
+      AD_FAIL();
+    }
+  };
+  auto computeQuadsForResultRow = [&resolveVariable](
+                                      auto& templates, auto& result,
+                                      const IdTable& idTable, uint64_t row,
+                                      const VariableToColumnMap&
+                                          variableColumns) {
+    for (const auto& [s, p, o, g] : templates) {
+      auto subject = resolveVariable(idTable, row, variableColumns, s);
+      auto predicate = resolveVariable(idTable, row, variableColumns, p);
+      auto object = resolveVariable(idTable, row, variableColumns, o);
+      auto graph = resolveVariable(idTable, row, variableColumns, g);
+
+      if (!subject.has_value() || !predicate.has_value() ||
+          !object.has_value() || !graph.has_value()) {
+        continue;
+      }
+      result.emplace_back(std::array{*subject, *predicate, *object, *graph});
+    }
+    return result;
+  };
+  std::vector<IdTriple<>> toInsert;
+  std::vector<IdTriple<>> toDelete;
+  // The maximum result size is size(query result) x num template rows. The
+  // actual result can be smaller if there are template rows with variables for
+  // which a result row does not have a value.
+  toInsert.reserve(res->idTable().size() * toInsertTemplates.size());
+  toDelete.reserve(res->idTable().size() * toDeleteTemplates.size());
+  for (const auto& [pair, range] :
+       ExportQueryExecutionTrees::getRowIndices(query._limitOffset, *res)) {
+    auto& idTable = pair.idTable_;
+    for (uint64_t i : range) {
+      computeQuadsForResultRow(toInsertTemplates, toInsert, idTable, i,
+                               qet.getVariableColumns());
+      cancellationHandle->throwIfCancelled();
+
+      computeQuadsForResultRow(toDeleteTemplates, toDelete, idTable, i,
+                               qet.getVariableColumns());
+      cancellationHandle->throwIfCancelled();
+    }
+  }
+
+  // TODO<qup42> use the actual DeltaTriples object
+  DeltaTriples deltaTriples = DeltaTriples{index_};
+  deltaTriples.insertTriples(cancellationHandle, std::move(toInsert));
+  deltaTriples.deleteTriples(cancellationHandle, std::move(toDelete));
+
+  co_return;
```

Comment:
I don't quite see why this has to be a coroutine.
Make it a normal function and move it to ExportQueryExeuctionTrees.
If you need coroutines for the server thread scheduling mechanism, then use a separate thin layber in `Server.cpp`.